### PR TITLE
Introduces the COCOMeanAveragePrecision metric

### PR DESCRIPTION
**Pull request by LukeWood**
COCO mean average precision measures the mean average precision using a bucketized computation approach.   This is done for tensorflow compatibility, performance, and to support multiple `update_state` calls.

Please note that the current metric is extremely slow despite my attempts to vectorize the computation.   I will be running a thorough performance benchmark using internal tooling following submission of this commit, then will re-visit and weight options.

### CODE REVIEW

**Comment by bhack** [comment]
> Please note that the current metric is extremely slow despite my attempts to vectorize the computation. I will be running a thorough performance benchmark using internal tooling following submission of this commit, then will re-visit and weight options.

Can we rely on XLA `jit_compile` in cases like this one?

File: keras_cv/metrics/coco/mean_average_precision.py [hunk 1]
```diff
@@ -0,0 +1,316 @@
...
+        self.true_positive_buckets.assign(tf.zeros_like(self.true_positive_buckets))
+        self.false_positive_buckets.assign(tf.zeros_like(self.false_positive_buckets))
+        self.ground_truths.assign(tf.zeros_like(self.ground_truths))
+
+    @tf.function()
+    def update_state(self, y_true, y_pred, sample_weight=None):
+        num_images = tf.shape(y_true)[0]
+
+        if sample_weight is not None:
+            raise ValueError("Received unsupported `sample_weight` to `update_state`")
```
Comment (fchollet) [inline keras_cv/metrics/coco/mean_average_precision.py:140]: This makes it sound like the sample_weight format is just invalid -- but in fact there are no valid values. Perhaps "{self.__class__.__name__} does not support `sample_weight`"

File: keras_cv/metrics/coco/mean_average_precision.py [hunk 2]
```diff
@@ -0,0 +1,316 @@
...
+
+        result = tf.reshape(
+            result.stack(), (self.num_class_ids, self.num_iou_thresholds)
+        )
+        result = tf.math.reduce_mean(result, axis=-1)
+        # tf.print('result', result)
+        result = tf.math.reduce_sum(result, axis=0) / tf.cast(
+            present_categories, tf.float32
+        )
+        # tf.print('result2', result)
```
Comment (fchollet) [inline keras_cv/metrics/coco/mean_average_precision.py:315]: Here too

File: keras_cv/metrics/coco/numerical_tests/mean_average_precision_test.py [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+import tensorflow as tf
+
+from keras_cv.metrics.coco import COCOMeanAveragePrecision
+from keras_cv.utils import bbox
+
+SAMPLE_FILE = os.path.dirname(os.path.abspath(__file__)) + "/sample_boxes.npz"
+
+
+class MeanAveragePrecisionTest(tf.test.TestCase):
+    """Unit tests that test Keras COCO metric results against the known good ones of
```
Comment (fchollet) [inline keras_cv/metrics/coco/numerical_tests/mean_average_precision_test.py:26]: Please fix docstring format

**Comment by LukeWood** [comment]
> > Please note that the current metric is extremely slow despite my attempts to vectorize the computation. I will be running a thorough performance benchmark using internal tooling following submission of this commit, then will re-visit and weight options.
> 
> Can we rely on XLA `jit_compile` in cases like this one?


There are dynamic dimensions in filter_bboxes :(.  Maybe I can reorganize the loops to do filtering first then call XLA compiled functions in the inner loops???  I will try this after I perform my benchmark.

File: keras_cv/metrics/coco/mean_average_precision.py [hunk 4]
```diff
@@ -0,0 +1,316 @@
...
+            over each incrementally selected set of bounding boxes, we instead
+            place them into buckets.  This makes distributed computation easier.
+            Increasing buckets improves accuracy of the metric, while decreasing
+            buckets improves performance.  This is a tradeoff you must weight
+            for your use case.  Defaults to 10,000 which is sufficiently large
+            for most use cases.
+
+    Usage:
+
+    COCOMeanAveragePrecision accepts two Tensors as input to it's `update_state` method.
```
Comment (fchollet) [inline keras_cv/metrics/coco/mean_average_precision.py:54]: Nit: method names should have `()` at the end when referring to them (to mark them as functions/methods)

File: keras_cv/metrics/coco/mean_average_precision.py [hunk 5]
```diff
@@ -0,0 +1,316 @@
...
+    These Tensors represent bounding boxes in `corners` format.  Utilities
+    to convert Tensors from `xywh` to `corners` format can be found in
+    `keras_cv.utils.bbox`.
+
+    Each image in a dataset may have a different number of bounding boxes,
+    both in the ground truth dataset and the prediction set.  In order to
+    account for this, you may either pass a `tf.RaggedTensor`, or pad Tensors
+    with `-1`s to indicate unused boxes.  A utility function to perform this
+    padding is available at `keras_cv_.utils.bbox.pad_bbox_batch_to_shape`.
```
Comment (fchollet) [inline keras_cv/metrics/coco/mean_average_precision.py:63]: ()

**Comment by bhack** [comment]
> There are dynamic dimensions in filter_bboxes :(. Maybe I can reorganize the loops to do filtering first then call XLA compiled functions in the inner loops??? I will try this after I perform my benchmark.

It could be interesting. As we have already discussed It in another ticket, I don't have acess to your internal tools so let me know when you have some numbers to share.

File: keras_cv/metrics/coco/numerical_tests/mean_average_precision_test.py [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+            iou_thresholds=[0.75],
+            max_detections=100,
+            area_range=(0, 1e5**2),
+        )
+
+        mean_average_precision.update_state(y_true, y_pred)
+        result = mean_average_precision.result().numpy()
+        self.assertAlmostEqual(result, 0.729, delta=0.06)
+
+    # TODO(lukewood): re-enable after performance testing
```
[the remark this answers is not in the document]
Comment (LukeWood) [reply keras_cv/metrics/coco/numerical_tests/mean_average_precision_test.py:103]: This is with respect to profiling.  I'd like to revisit this a bit later as the `large images` case also has the highest error margin in COCORecall which I find very curious.
